### PR TITLE
Feature/smelter enhancements

### LINKED ISSUE #1963: Proposed Smelter Enhancement
Proposed enhancement for the Smelter Worker :)

When deconstructing enchanted weapons/tools/armour - give him a chance to salvage the enchant as an enchanted book, based on his hut level.

Hut Level 0/1 = NO chance
Hut Level 2 = 10%
Hut Level 3 = 25%
Hut Level 4 = 40%
Hut Level 5 = 60%

### PR DESCRIPTION
**Pull request by TheYoxy**
Closes #1963 

# Changes proposed in this pull request:
- Added a chance to loot an enchanted book with the enchantments of this tool/weapon/gear.
This chance look about the durability of the item. If it's low, you have a lower chance to loot a book.

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java [hunk 1]
```diff
@@ -150,11 +163,19 @@ private AIState smeltStuff()
...
             final ItemStack material = materialTuple.getFirst();
-            if(!ItemStackUtils.isEmpty(material))
+            if (!ItemStackUtils.isEmpty(material))
             {
                 material.setCount(materialTuple.getSecond());
                 material.setItemDamage(0);
                 new InvWrapper(worker.getInventoryCitizen()).setStackInSlot(slot, material);
+                if (stack.isItemEnchanted())
+                {
+                    if (ENCHANTED_BOOK_CHANCE[getOwnBuilding().getBuildingLevel() - 1] < new Random().nextInt(100))
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java]: just to make sure, better check > 0 here.

**Review by Raycoms — COMMENTED** [review]
Looks great, I'd just add one security check but the rest looks good =)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java [hunk 2]
```diff
@@ -150,11 +163,19 @@ private AIState smeltStuff()
...
+                    {
+                        final ItemStack book = extractEnchantFromItem(stack);
+                        new InvWrapper(worker.getInventoryCitizen()).insertItem(InventoryUtils.findFirstSlotInItemHandlerWith(new InvWrapper(worker.getInventoryCitizen()),ItemStack::isEmpty),book,false);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java]: MAJOR Split this 203 characters long line (which is greater than 180 authorized). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java [hunk 3]
```diff
@@ -89,15 +101,16 @@ public EntityAIWorkSmelter(@NotNull final JobSmelter job)
...
+                                  + INTELLIGENCE_MULTIPLIER * worker.getCitizenData().getIntelligence());
         worker.setCanPickUpLoot(true);
     }
 
     /**
      * He will smelt down armor, weapons and tools to smaller pieces here.
+     *
      * @return the next state to go to.
      */
     private AIState smeltStuff()
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java]: MAJOR Refactor this method to reduce its Cognitive Complexity from 22 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java (hunk 1 above)
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1066)
MAJOR Assign this magic number 100 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 52 issues
* ![MAJOR][MAJOR] 52 major

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [init#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/init#): Add a 'package-info.java' file to document the 'init' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [locations#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/locations#): Add a 'package-info.java' file to document the 'locations' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [management#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/management#): Add a 'package-info.java' file to document the 'management' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 51 issues
* ![MAJOR][MAJOR] 51 major

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [init#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/init#): Add a 'package-info.java' file to document the 'init' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [locations#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/locations#): Add a 'package-info.java' file to document the 'locations' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [management#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/management#): Add a 'package-info.java' file to document the 'management' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [handlers#](https://github.com/Minecolonies/minecolonies/blob/ca8ed7f154f66f227028d61a7958a9074223fa3c/src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers#): Add a 'package-info.java' file to document the 'handlers' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java]: definitely as well.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java (hunk 3 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java]: That's okay, you can leave this.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java (hunk 2 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/smelter/EntityAIWorkSmelter.java]: make a line break here on new InvWrapper.

**Comment by Raycoms** [comment]
Looks good :shipit: 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/2492/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 47 issues
* ![MAJOR][MAJOR] 47 major

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [init#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/colony/requestsystem/init#): Add a 'package-info.java' file to document the 'init' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [locations#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/colony/requestsystem/locations#): Add a 'package-info.java' file to document the 'locations' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [management#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/colony/requestsystem/management#): Add a 'package-info.java' file to document the 'management' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [handlers#](https://github.com/Minecolonies/minecolonies/blob/22ffd20a4f392bc23bb44b9d9a349f926c3f24fa/src/main/java/com/minecolonies/coremod/colony/requestsystem/management/handlers#): Add a 'package-info.java' file to document the 'handlers' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**